Dictifying to any type to allow multiple functions

francispoulin opened the pull request.

tomchor reviewed (approved):
Provided that you've already tested this, it looks good to me!

francispoulin commented:
> Provided that you've already tested this, it looks good to me!

Thanks @tomchor .  It seems that some regression testes are failing.  

Is this a concern @ali-ramadhan ?

tomchor commented:
Yeah, I've been noticing that a lot of simple PRs have been having failing tests lately. I haven't had time to investigate but I do find it kinda odd...

ali-ramadhan commented:
@francispoulin This particular failure happens occasionally and I think it's related to #1179 but haven't really investigated. Usually I just restart the test and it passes the second time. Did it here so hopefully it should pass.

@tomchor Doc builds seem to be hanging recently across most PRs and commits. I haven't had time to investigate but hope to soon...

glwagner commented on src/OutputWriters/netcdf_output_writer.jl:
```diff
@@ -13,7 +13,7 @@ using Oceananigans.Diagnostics: WindowedSpatialAverage
 
 dictify(outputs) = outputs
-dictify(outputs::NamedTuple) = Dict(string(k) => dictify(v) for (k, v) in zip(keys(outputs), values(outputs)))
+dictify(outputs::NamedTuple) = Dict{Any,Any}(string(k) => dictify(v) for (k, v) in zip(keys(outputs), values(outputs)))
```
What issue is this solving?

ali-ramadhan replied:
Not sure why the issue pops up since we test for it but when @francispoulin tried to save two function outputs, e.g.

```
(output1 = function1, output2 = function2)
```

`dictify` complained that it couldn't convert `typeof(function2)` to `typeof(function1)` so I suspected that it created a `Dict{String,typeof(function1)}` for some reason.

Explicitly creating a `Dict{Any,Any}` fixes this issue as expected but I don't see why it's necessary...

glwagner replied:
Are we sure there wasn't a bug in the code? If we test it then it should work? Or we can create another test that fails...

ali-ramadhan replied:
We looked at it but it looked exactly like what we were doing: https://github.com/CliMA/Oceananigans.jl/blob/ff1131c8ec5f80d563bae59f7967dc0c72a2ebc6/test/test_netcdf_output_writer.jl#L335-L359

Agree that we shouldn't blindly put band-aids in the code. We could keep this PR open until we can put in a proper fix + test.

francispoulin commented:
The purpose of this PR was to save mutliple scalars using `NetCDFOutputWriter` and there seemed to be some problems, i.e. some checks fail.

@glwagner mentioned that maybe `using Statistics: norm` might work without any additional work.  Is this the case?  If yes then I am happy to close this PR and give it a try on the shallow water Bickley jet example.

glwagner commented:
> The purpose of this PR was to save mutliple scalars using `NetCDFOutputWriter` and there seemed to be some problems, i.e. some checks fail.
> 
> @glwagner mentioned that maybe `using Statistics: norm` might work without any additional work. Is this the case? If yes then I am happy to close this PR and give it a try on the shallow water Bickley jet example.

Ah... the change is fairly innocuous, there's just a few things that make me nervous. 1) It _shouldn't_ fix the problem that it's intended to fix. Unless there's something I am missing... if it fixes the problem that was raised, it's due to some unintended side effect that we don't understand? (Providing some detail and the error message that was received could help, or it could be raised in an issue). A more grave concern is 2) it will slow down the code and cause memory allocation. This is almost always completely negligible since other parts of the code matter more, but it could affect small models, our CI pipeline, etc. It's hard to intentionally throw away type stability :-/ So in a way it's creating technical debt and possibly creating work for us in the future...

francispoulin commented:
Thanks @glwagner.  I certainly don't want to make any changes that would either slow down code or make things harder to maintain.  I can't say that I know this is the right direction to go but let me try and articulate what I want to do and maybe you can suggest a direction to proceed?

The basic idea can be found by looking at `shallow_water_Bickley_jet.jl`.  In this example we define a `perturbation_norm` in line 134 that computes the norm of `v`.  Then in line 158 we define an output_writer to output this norm.  There is nothing special about v and I might want to use use `u` as well.  To do this requires two steps:

1. Defining a new analogous function for the perturbation norm of `u`.  I'm hoping we can avoid these functions completely with `using Statistics: norm` but I don't know if that's the case.
2. In line 161 we need to list these two norms.  When I tried that there was an error message, and hence what inspired this PR.

I don't think this PR needs to happen and there are probably better ways of addressing points 1 and 2.  Any advice?

glwagner commented:
I think you might want to use a callable object so that you can reuse `norm` (note that `norm` does work on `Field` now --- but this isn't the central issue). For example:

```julia
struct FieldNorm{F}
    field :: F
end

function (field_norm::FieldNorm)(model)
    compute!(field_norm.field)
    return norm(field_norm.field)
end
```

Now you can re-use this object for multiple outputs:

```julia
outputs = (u_norm = FieldNorm(u), v_norm = FieldNorm(v))
```

The way the example is currently written, the field `v` is "captured" from global scope by the function `perturbation_norm`. Using a the global scope in this way is convenient if you only need to use the code inside the function once, but it can't be reused as you found. In order to reuse a function, you need to capture the data in a different way. Using a callable object to capture the data you want to use for function output achieves this.

francispoulin commented:
Thanks @glwagner , that is a cleaner approach and I will try it but I don't think that's going to help me write multiple fields, but I hope to be wrong.  I will let you know.

glwagner commented:
Is this PR needed anymore?

francispoulin commented:
I don't believe so and will therefore close it.
